test/test_torch.py

```diff
@@ -4873,6 +4873,128 @@ def test_diff(self, device, dtype):
+                actual[i] = torch.nan_to_num(actual[i], posinf=0, neginf=0) if dtype != torch.complex64 else actual[i]
+            return actual, expected
+
+        # shape, dims format
+        test_cases = (
+            ((5,), (0,)),
+            ((4, 4), (0, 1)),
+            ((3, 3, 3), (-1, 0)),
+            ((4, 4, 4), (2,)),
+            ((4, 4, 4), (0, 1)),
+            ((4, 4, 4, 3), (0, 2, 3)),
+            ((4, 5, 2, 4, 3), (1, 2)),
+            ((4, 3, 6, 5, 3), (2, 4)),
+            ((4, 3, 2, 5, 3), (0, 1, 2, 3, 4)),
+        )
+
+        for case, space_fn in product(test_cases, (create_scalar, create_list, create_tensors)):
+            for contig in [True, False]:
+                shape, dims = case
+
+                # filter shape by dims before passing filtered shape to create_* functions
+                filtered_shape = filter_shape(shape, dims)
+
+                spacing = space_fn(filtered_shape)
+                t = make_tensor(shape, device=device, dtype=dtype, noncontiguous=not contig)
+                t_np = t.cpu().numpy()
+
+                actual = torch.gradient(t, spacing=spacing, dim=dims, edge_order=1)
+                if space_fn == create_tensors and spacing[0].device != 'cpu':
+                    spacing = [space.cpu().detach().numpy() for space in spacing]
+                expected = np.gradient(t_np, *self._wrap_to_list(spacing), axis=dims, edge_order=1)
+                actual, expected = inf_nan_preprocess(list(actual), self._wrap_to_list(expected))
+                self.assertEqual(actual, self._wrap_to_list(expected), equal_nan="relaxed")
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
+    def test_gradient_extreme_cases(self, device, dtype):
+        # Test behaviour for inf and nan values
+        actual = torch.gradient(torch.tensor([2, -2, inf, inf, -inf, -inf, inf, 3, -inf, 2, nan, nan, 3, inf, nan]))
+        expected = np.gradient(np.array([2, -2, inf, inf, -inf, -inf, inf, 3, -inf, 2, nan, nan, 3, inf, nan]))
+        self.assertEqual(actual, self._wrap_to_list(expected),)
+
+        # Test behaviour in very big tensors
+        large_size = 100000
+        t = make_tensor((large_size,), device, dtype)
+        t_np = t.cpu().numpy()
+        coordinates_np = list(np.random.randn(large_size))
+        coordinates = [torch.tensor(coordinates_np, device=device)]
+        actual = torch.gradient(t, spacing=coordinates, dim=0, edge_order=1)
+        expected = [np.gradient(t_np, coordinates_np, axis=0, edge_order=1)]
+        self.assertEqual(actual, expected)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
+    def test_gradient_type_promotion(self, device, dtype):
+        # Test type promotion
+        inputs = (torch.randn(4, 4), torch.randn(4, 4, dtype=torch.complex64))
+
+        spacing = (
+            3, 3., complex(3, .5), [1, 2], [1.3, .7], (torch.randn(4), torch.randn(4))
```

Comment:
same make_tensor()/device issue here.

What about a list of complex scalars, too?

Reply:
tensor coordingates, complex64, int64.. right now we have just float.